### PR TITLE
new Go support function 'buildGoPackage'

### PR DESCRIPTION
**Pull request by ghost**
This PR is for a new language-specific build funtion for Go.

I have some practical package expressions that use this fuction that I can append to the PR as well.
Also, I have a script that spits out basic expressions for building a Go package using only a Go import line, but I'm going to wait to do some more testing first.

This will take care of https://github.com/NixOS/nixpkgs/issues/3080.

### CODE REVIEW

File: pkgs/development/go-modules/generic/default.nix [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    local GOPATH="$NIX_BUILD_TOP/go:$GOPATH"
+
+    if [ -n "$subPackages" ] ; then
+	for p in $subPackages ; do
+            go install -p $NIX_BUILD_CORES -v $packagePath/$p
+	done
+    else
+	find . -type d | while read d; do
+            for i in $d/*.go; do
+                go install -p $NIX_BUILD_CORES -v $d
```
Comment (lethalman) [inline pkgs/development/go-modules/generic/default.nix]: This is executed for every .go file, why?

File: pkgs/development/go-modules/generic/default.nix [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  '';
+
+  checkPhase = args.checkPhase or ''
+    runHook preCheck
+
+    local GOPATH="$NIX_BUILD_TOP/go:$GOPATH"
+
+    if [ -n "$subPackages" ] ; then
+	for p in $subPackages ; do
+            go install -p $NIX_BUILD_CORES -v $packagePath/$p
```
Comment (lethalman) [inline pkgs/development/go-modules/generic/default.nix]: Shouldn't this be go test?

File: pkgs/top-level/go-packages.nix [hunk 3]
```diff
@@ -0,0 +1,55 @@
+/* This file defines the composition for Go packages. */
+
+{ pkgs, overrides, stdenv }:
+
+let self = _self // overrides; _self = with self; {
```
Comment (lethalman) [inline pkgs/top-level/go-packages.nix]: This set must have also a `go` attribute, to know which go version are we using for this set of packages.

**Comment by lethalman** [comment]
I'm willing to review this, @ehmry would you fix the couple of notes I've added to the changeset?

File: pkgs/top-level/go-packages.nix [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  inherit (pkgs) buildGoPackage fetchgit fetchhg fetchurl;
```
Comment (lethalman) [inline pkgs/top-level/go-packages.nix]: These must be all inputs. Don't inherit from pkgs, thanks.

**Comment by ehmry** [comment]
@lethalman I've fixed the problems you pointed out but go.crypto doesn't build anymore, which breaks anything I wanted to do. Doesn't work with Go 1.3 or 1.4

The amount of work that would be involved in keeping buildGoPackage working for different package layouts and the quirks of go build/install is frightening me into considering calling Go too fragile and closing the PR.

All things considered I would prefer to build our own version of 'go build' that reads everything it needs from a builder environment and resolves and installs the way we like, but that not might be practical.

**Comment by ehmry** [comment]
A lot of go build is exported as public interfaces, so it might not be so insane.
https://github.com/golang/go/tree/master/src/cmd/go

**Comment by lethalman** [comment]
I'm sorry I can't say why it breaks crypto.

**Comment by lethalman** [comment]
It's that I don't get why the changes I proposed hurt the build of crypto. Will have to check locally.

**Comment by ehmry** [comment]
Bah

**Comment by lethalman** [comment]
Btw the problem @ehmry seems more like fetchhg not being deterministic.

**Comment by lethalman** [comment]
I was able to successfully build `goPackages.net`. @ehmry I don't understand the reaction, are you going to drop the whole work or perhaps open a new PR? Didn't get it.

**Comment by lethalman** [comment]
I was even able to build `go13Packages.net`. This is a quick diff for having go14Packages and go13Packages:

```
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -7594,14 +7594,24 @@ let
 
   ### DEVELOPMENT / GO MODULES
 
-  buildGoPackage = import ../development/go-modules/generic {
-    inherit go;
-  };
+  go13Packages = recurseIntoAttrs (callPackage ./go-packages.nix {
+    go = go_1_3;
+    buildGoPackage = import ../development/go-modules/generic {
+      go = go_1_3;
+    };
+    overrides = (config.goPackageOverrides or (p: {})) pkgs;
+  });
 
-  goPackages = recurseIntoAttrs (callPackage ./go-packages.nix {
+  go14Packages = recurseIntoAttrs (callPackage ./go-packages.nix {
+    go = go_1_4;
+    buildGoPackage = import ../development/go-modules/generic {
+      go = go_1_4;
+    };
     overrides = (config.goPackageOverrides or (p: {})) pkgs;
   });
 
+  goPackages = go14Packages;
+
   ### DEVELOPMENT / LISP MODULES
 
   asdf = callPackage ../development/lisp-modules/asdf {
```

Will try this PR for a couple of real projects.

**Comment by ehmry** [comment]
I'm going to drop the work, but if you want to take it on I'm willing to help.

**Comment by lethalman** [comment]
Ok so I'm officially working on this. FYI @iElectric I'm working on porting ngrok to this new system.

**Comment by iElectric** [comment]
:Beers:

**Pull request closed without merging**